Auth group -> Aerie role mapping

skovati opened the pull request:
Closes https://github.com/NASA-AMMOS/aerie/issues/1279

This PR introduces a few generic helper functions that map auth provider groups to Aerie roles. The mapping is proved by a new env var `AUTH_GROUP_ROLE_MAPPINGS`, which is a stringified JSON object that maps auth group (keys) to Aerie roles (values), e.g.

```
{
  "group_A": ["aerie_admin", "user"],
  "group_B": ["viewer"]
}
```

This allows external auth providers (Auth0, AWS Cognito) to handle group membership and be the source of truth, but map them to Aerie roles when SSO is used.

Additionally, this PR updates the `DEFAULT_ROLE` env var to be a list. This allows listing multiple default roles by priority. The following issue goes into more depth with this priority algorithm: https://github.com/NASA-AMMOS/aerie/issues/1279

Pen-ultimately, this PR updates the build process to use `node` `v20.*`, which is what the Gateway runs with inside the `aerie-gateway` container.

Finally, this PR brings tests to the Gateway!

Docs are WIP

skovati committed:
make `DEFAULT_ROLE` env var a string array

skovati committed:
add new `AUTH_GROUP_ROLE_MAPPINGS` env var

skovati committed:
Change `GroupRoleMapping` type to be a dynamically keyed obj

skovati committed:
add group -> role validation function

skovati committed:
Add helper function to get default role for auth group

skovati committed:
add helper functions to map auth groups to default / allowed roles

skovati committed:
extract LDAP groups from CAM response, map to roles

skovati committed:
rename function for clarity

skovati committed:
refactor for clarity

skovati committed:
add tests

skovati committed:
fix lint error

skovati committed:
fix destructuring names

Mythicaeda commented on docs/ENVIRONMENT.md:
```diff
@@ -10,7 +10,7 @@ This document provides detailed information about environment variables for the
 | `AUTH_UI_URL`               | URL of Auth provider's login UI. Returned to the UI if SSO token is invalid, so user is redirected   | `string` | https://atb-ocio-12b.jpl.nasa.gov:8443/cam-ui  |
 | `AUTH_SSO_TOKEN_NAME`       | The name of the SSO tokens the Gateway should parse cookies for. Likely found in auth provider docs. | `array`  | ["iPlanetDirectoryPro"]                        |
-| `DEFAULT_ROLE`              | Default role when authentication is enabled.                                                         | `array`  | user                                           |
+| `DEFAULT_ROLE`              | Default roles when authentication is enabled.                                                        | `array`  | ["user"]                                       |
```
`DEFAULT_ROLE`s description should either explain the algorithm or point to docs that do so. Since the full docs are still WIP, probably the latter once they're finished.

skovati replied:
+1, will update once docs are complete.

skovati replied:
Linked to docs for now, which are still WIP.

Mythicaeda commented on docs/ENVIRONMENT.md:
Docs are missing for `AUTH_GROUP_ROLE_MAPPINGS`

skovati replied:
Added a small blurp and a link to where the authn docs will exist.

Mythicaeda commented on src/packages/auth/types.ts:
```diff
@@ -50,3 +50,10 @@ export interface AuthAdapter {
   logout(req: Request): Promise<boolean>;
 }
+
+export type GroupRoleMapping = {[key: string] : string[]}
```
The `[key: string]` syntax is used to say that the key of this mapping is any string, right?

skovati replied:
Correct. It enforces the `string: string[]` type but doesn't specify specific keys. This way, keys can just be the auth groups, allowing for quick lookups.

Mythicaeda commented on src/env.ts:
```diff
@@ -63,7 +66,23 @@ function parseArray<T = string>(value: string | undefined, defaultValue: T[]): T
+/**
+ * Parses a JSON env var string into a GroupRoleMapping object, which has dynamically named keys
+ */
+function parseGroupRoleMappings(value: string | undefined, defaultValue: GroupRoleMapping): GroupRoleMapping {
+  if (typeof value === 'string') {
+    try {
+      return JSON.parse(value);
+    } catch (e) {
+      console.error(e);
       return defaultValue;
     }
   }
```
For GroupRoleMappings, I'm not sure if it makes sense to return the default value if parsing fails.

skovati replied:
Instead just returning a hardcoded `{}`?

Mythicaeda replied:
That or calling a parse fail a fatal error

skovati replied:
Change has been made to make failed json parse a fatal error.

Mythicaeda commented on src/packages/auth/functions.ts:
```diff
@@ -134,7 +134,7 @@ export async function login(username: string, password: string): Promise<AuthRes
         };
       } else {
-        const { allowed_roles, default_role } = await getUserRoles(username, DEFAULT_ROLE, ALLOWED_ROLES);
+        const { allowed_roles, default_role } = await getUserRoles(username, DEFAULT_ROLE[0], ALLOWED_ROLES);
```
This behavior looks incorrect, as it is a CAM-based login ignoring the LDAP Group roles. Are they not available in this response?

skovati replied:
I had initially left this new mapping logic out of the user / pass based login flow. I'll work on adding it now.

skovati replied:
Change has been made. Basically just extracted the sso token from our user / pass login, and then called `ssoLogin` using that token, so that we hit the code path that handles this group -> role mapping logic.

Mythicaeda commented on src/packages/auth/functions.ts:
```diff
@@ -178,3 +178,77 @@ export async function session(authorizationHeader: string | undefined): Promise<
+export function mapGroupsToRoles(groupList: string[]): UserRoles {
+  const { DEFAULT_ROLE, ALLOWED_ROLES, AUTH_GROUP_ROLE_MAPPINGS } = getEnv();
+
+  let default_role = DEFAULT_ROLE[0];
+  let allowed_roles  = ALLOWED_ROLES;
```
Is this initialization here to cover the "no mappings" case?

skovati replied:
Yep, just using a mutable var instead of returning within two branches. I might refactor this for clarity though.

skovati replied:
Refactored. Let me know if this is more clear now.

Mythicaeda replied:
It is, thanks!

Mythicaeda commented on src/packages/auth/functions.ts:
```diff
@@ -178,3 +178,77 @@ export async function session(authorizationHeader: string | undefined): Promise<
+  };
+}
+
+export function getGroupsWithMappings(authGroups: string[]): string[] {
+  const { AUTH_GROUP_ROLE_MAPPINGS } = getEnv();
+  const authGroupsSet = new Set(authGroups);
+
+  return Object
+    .keys(AUTH_GROUP_ROLE_MAPPINGS)
+    .filter(mappedGroup => authGroupsSet.has(mappedGroup));
+}
```
Minor, nonblocking: could you please comment this function? It took a couple readthroughs to realize that this was taking in a set of groups and removing the ones that Aerie doesn't have role mappings for. Either a one-liner at the top or inlines like you did for `getAllAllowedRolesForAuthGroups` would be sufficient.

skovati replied:
Added a few function docs!

Mythicaeda commented on test/auth.test.ts:
```diff
@@ -0,0 +1,123 @@
+  describe(
+    'reference example', () => {
+
+      const default_roles = ['viewer', "user", "aerie_admin"];
```
Minor, style: mixed quote styles in use here and across the file

skovati replied:
Good catch, I'll add a format check to CI as well.

skovati committed:
add format check to CI

skovati committed:
update gateway to node 20

skovati committed:
run prettier

skovati committed:
refactor `mapGroupsToRoles` for clarity

Mythicaeda commented on src/packages/auth/functions.ts:
```diff
@@ -212,37 +220,37 @@ export function getDefaultRoleForAllowedRoles(allowedRoles: string[]): string {
+      return {
+        allowed_roles,
+        default_role: getDefaultRoleForAllowedRoles(allowed_roles),
+      }
```
Return (or error) is missing for the `else` case of this if

skovati replied:
If `mappedGroupMembership` is empty, allowed roles and default roles will also be empty. I think we can handle this error at a broader level (i.e. `loginSSO`), and return a failed login if `allowed_roles` is empty after checking the DB.

skovati replied:
Although that does create an issue of "stale" membership, e.g. LDAP membership might change external to Aerie, but our DB is the source of truth once a user is created.

skovati replied:
Maybe we can just take the intersection of DB `allowed_roles` and auth group mapped `allowed_roles`?

Mythicaeda replied:
If the user signing in isn't new, the DB check will pass, because it'll be using the DB data, not the empty data. Additionally, the issue here is that the method isn't explicitly returning empty, it's returning nothing in the else case

Mythicaeda replied:
RE: Taking the intersection/updating the DB to reflect changes in LDAP, while I'm not against it, I'd say check we should check in with @dandelany and/or @joswig, as that's an additional behavior change.

skovati replied:
Ah wrote those comments based on local changes I had where I removed the `length > 0` check entirely 🤦

That way, we'll forward an empty `allowed_roles` list if `mappedGroupMembership` is empty, and the downstream auth adapter can handle that case however they want (e.g. continue to check user in DB, return failed login, etc)

skovati committed:
Extract SSO token and use `loginSSO` in CAM user / pass flow

skovati committed:
make failed group mapping JSON parse a fatal error

skovati committed:
Conditionally update roles in DB with auth group mapping roles

skovati committed:
fail login if `allowed_roles` is empty

skovati committed:
run prettier

skovati committed:
update SSO auth env var docs

skovati committed:
add `getGroupsWithMappings` docs

dandelany commented:
I was able to test this locally & get it working with some help with @skovati ... testing steps for anyone following along:
* Pull this branch locally and run `npm install && npm run build`
* In the `aerie` repo, modify your docker compose file to add this to the `aerie_gateway` section:
```
aerie_gateway:
    build:
        context: ../aerie-gateway
```
* also in docker compose, comment out the line `# image: "ghcr.io/nasa-ammos/aerie-gateway:develop"`
* Then, to test with LDAP groups, add this to the `aerie_gateway: environment: ` section of the docker compose, where "your_ldap_group" is a group you have membership in
```
  DEFAULT_ROLE: |
    ["aerie_admin"]
  AUTH_GROUP_ROLE_MAPPINGS: |
    {
      "your_ldap_group": [ "user", "viewer" ]
    }
```
* Also change the `AUTH_TYPE` variable to `AUTH_TYPE: cam`
* Then `docker compose up -d --build` to run Aerie with these settings

This worked for me locally 👍 The only remaining question/work is related to what happens when the user’s LDAP group(s)/roles change in the middle of a session. It would be nice if Aerie handled this gracefully by eg. automatically changing your roles or logging you out & making you login again. But the default requirement is that the LDAP mapping is only checked on Login & is valid for your entire Aerie session. @skovati is working on finalizing this behavior & will update when done.

Thanks!

skovati committed:
only upsert roles on first session login

skovati committed:
check JWT validity when determining if session is new

Mythicaeda commented on src/packages/auth/adapters/CAMAuthAdapter.ts:
```diff
@@ -94,14 +103,35 @@ async function loginSSO(ssoToken: any): Promise<AuthResponse> {
+    // We only do this when the user is logging in for the first time
+    // in a session (user cookie DNE) for DB performance reasons,
+    // and to avoid roles changing underneath the user during a session
+    const isNewUserSession = !(userCookie && cookieIsValid(userCookie));
```
In the SSO case, the user token could still be valid but the user has to relog due to SSO session expiry

Mythicaeda replied:
Might be a UI bug, since the user token wasn't invalidated even when manually logging out.

skovati replied:
I've significantly reworked this section. The conditional logic for upserting roles to the DB now depends on diffing the freshly mapped allowed roles and the existing DB allowed roles for this user.

This entirely removes the "session" concept; users are auth'd and roles are updated exactly when we detect, not just once the first time users log in. This significantly simplifies upstream (i.e. the UI), since we don't have to worry about clearing user tokens there to get a new "session", instead the a Gateway will always provide a JWT that matches the allowed roles in the DB.

I've done some testing, and the UI handles roles changing underneath it quite well. I only see one small UI change (resetting active role cookie to default if it's no longer included in your new allowed roles) that accounts for the edge case of losing a role while using the UI.

Mythicaeda commented on src/packages/auth/functions.ts:
```diff
@@ -46,28 +55,53 @@ export async function getUserRoles(
   }
 }
 
+export async function syncRolesToDB(username: string, default_role: string, allowed_roles: string[]) {
+  await deleteUserAllowedRoles(username);
+  await upsertUserRoles(username, default_role, allowed_roles);
+}
+
```
Nonblocking but would _strongly_ prefer if these were in a single transaction instead of 2.

skovati replied:
Agreed, they originally were, but I removed them while debugging and never restored them

skovati replied:
Change has been made.

skovati committed:
wrap role updates in transaction

skovati committed:
Change upsert conditional logic

Upserts DB roles when they have changed, if group -> role mappings are
the source of truth. This removes the idea of a user "session", and auth
is reconsidered during each request, which simplifies things upstream.

Mythicaeda reviewed (approved):
LGTM!
